### PR TITLE
Add dtype validation

### PR DESCRIPTION
**Pull request by mmoscickiTT**
### PR Category
Feature

### Summary
binary_ng eltwise binary ops previously accepted operand dtype combinations that the kernels do not actually support, for example mixed float32/bfloat16, mixed integer widths, or unsupported dtypes like uint8 on arithmetic ops. Those cases could fail late, produce incorrect results, or behave inconsistently depending on the op and hardware path.

This PR adds an explicit per-op dtype policy for binary_ng and enforces it at runtime with TT_FATAL. Unsupported input dtypes and invalid mixed-dtype operand pairs now fail. A small set of intentional exceptions is preserved: mixed block-float tile dtypes (bfloat16 / bfloat8_b / bfloat4_b) on supported FPU ops, asymmetric quant/dequant/requant operands, and isclose float32↔bfloat16.

Tests and a few call sites (sampling logits, tutorial, broadcast tests) are updated to align operand dtypes before calling binary ops.

Relates to #33647

### Notes for reviewers

1. Call sites: I fixed the failures found in sanity runs (tt_log_probs.py, tutorial, broadcast/pow tests), but there may be other model or test code paths that still pass mismatched dtypes and will now TT_FATAL.

2. Mixed-dtype policy: The rules are split across binary_op_dtype_policy (per-op allowed dtypes) and is_dtype_combination_supported (operand-pair exceptions). Please sanity-check that the intentional exceptions match kernel behavior:

    Same dtype required for most ops
    Mixed bfloat16 / bfloat8_b / bfloat4_b allowed only on selected FPU ops
    quant / dequant / requant: tensor A per policy, tensor B must be float32
    isclose: only float32 ↔ bfloat16
    If any exception looks wrong or too narrow/broad, flag it.

3. Policy maintenance for new ops: Every BinaryOpType routed through binary_ng must be explicitly listed in binary_op_dtype_policy.cpp (and in supports_mixed_bfloat_tile_inputs / is_dtype_combination_supported if it has special mixed-dtype rules). A new op added to the enum but omitted from the policy switch will hit default: TT_THROW and fail at runtime. There is no permissive fallback.

    We do not yet have an exhaustivity test for this - happy to add that here or in a follow-up if reviewers think it belongs in this PR.

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:mmoscickiTT_33647_dtype_validation)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:mmoscickiTT_33647_dtype_validation)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:mmoscickiTT_33647_dtype_validation)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:mmoscickiTT_33647_dtype_validation)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:mmoscickiTT_33647_dtype_validation)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:mmoscickiTT_33647_dtype_validation)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:mmoscickiTT_33647_dtype_validation)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:mmoscickiTT_33647_dtype_validation)

### CODE REVIEW

**Review by Aswinmcw — APPROVED** [review]

**Review by mateusznowakTT — APPROVED** [review]

File: ttnn/tutorials/ttnn_basic_operations.ipynb [hunk 1]
```diff
@@ -141,7 +141,7 @@
     ")\n",
     "\n",
     "t5 = np.array([[5, 6], [7, 8]], dtype=np.float32).repeat(16, axis=0).repeat(16, axis=1)\n",
-    "tt_t5 = ttnn.Tensor(t5, device=device, layout=ttnn.TILE_LAYOUT)"
+    "tt_t5 = to_tt_tile(torch.from_numpy(t5))"
```
Comment (bbradelTT) [inline ttnn/tutorials/ttnn_basic_operations.ipynb]: What is the problem with ttnn.Tensor that makes you create this change? Does this need to be addressed [in a future issue/PR]?

**Review by bbradelTT — APPROVED** [review]

File: ttnn/tutorials/ttnn_basic_operations.ipynb (hunk 1 above)
Comment (mmoscickiTT) [reply ttnn/tutorials/ttnn_basic_operations.ipynb]: There’s no issue with ttnn.Tensor itself. The old code created tt_t5 as float32 from numpy while tt_t1 is bfloat16 via to_tt_tile helper function. After the PR’s stricter binary dtype validation, ttnn.mul(tt_t1, tt_t5) would TT_FATAL on mixed dtypes. I switched tt_t5 to to_tt_tile(...) so both operands are bfloat16, consistent with the rest of the tutorial.

**Review by tchedaTT — APPROVED** [review]

**Comment by KalaivaniMCW** [comment]
@mmoscickiTT FYI, I have triggered the Single card pipeline 
https://github.com/tenstorrent/tt-metal/actions/runs/27593138562

Blocking `binary_op(bf16_tensor, fp32_tensor)` for arithmetic binary ops might break some models. 
(bf16, fp32) mixed dtype should be working fine on main since both are floating point dtypes, ( but output takes input_a's dtype) 
Only difference might be whether it runs on FPU / SFPU 
Any failures you observed for this case ?

**Comment by mmoscickiTT** [comment]
> @mmoscickiTT FYI, I have triggered the Single card pipeline https://github.com/tenstorrent/tt-metal/actions/runs/27593138562
> 
> Blocking `binary_op(bf16_tensor, fp32_tensor)` for arithmetic binary ops might break some models. (bf16, fp32) mixed dtype should be working fine on main since both are floating point dtypes, ( but output takes input_a's dtype) Only difference might be whether it runs on FPU / SFPU Any failures you observed for this case ?

I have not seen failures for this in All Model Tests [1,2,3]. But I have seen some tests fail with this hence the changes in tt_log_probs, test_binary_bcast.py and the tutorial. In the pipeline you started I can unfortunetely see some models failing.

**Comment by mmoscickiTT** [comment]
> @mmoscickiTT FYI, I have triggered the Single card pipeline https://github.com/tenstorrent/tt-metal/actions/runs/27593138562
> 
> Blocking `binary_op(bf16_tensor, fp32_tensor)` for arithmetic binary ops might break some models. (bf16, fp32) mixed dtype should be working fine on main since both are floating point dtypes, ( but output takes input_a's dtype) Only difference might be whether it runs on FPU / SFPU Any failures you observed for this case ?

@KalaivaniMCW, looks like bf16xfp32 should stay supported, do you agree?

**Comment by mmoscickiTT** [comment]
> > @mmoscickiTT FYI, I have triggered the Single card pipeline https://github.com/tenstorrent/tt-metal/actions/runs/27593138562
> > Blocking `binary_op(bf16_tensor, fp32_tensor)` for arithmetic binary ops might break some models. (bf16, fp32) mixed dtype should be working fine on main since both are floating point dtypes, ( but output takes input_a's dtype) Only difference might be whether it runs on FPU / SFPU Any failures you observed for this case ?
> 
> @KalaivaniMCW, looks like bf16xfp32 should stay supported, do you agree?

I have updated the branch with fp32xbf16 support, re-running single card pipeline

**Comment by mmoscickiTT** [comment]
I have now run sanity, bpc, all model, and models extended.

| Test / job | Pipeline | Classification | Reason |
|---|---|---|---|
| Qwen2.5-32B unit [wh_llmbox_perf] | [all-models](https://github.com/tenstorrent/tt-metal/actions/runs/28088002561) | Flaky / infra | Job-level timeout; tests passed PCC 0.9998+. New to branch matrix, no main twin |
| Llama 3.1-8B data-parallel e2e (Galaxy) [wh_galaxy_perf] | [all-models](models) | Infra | Runner shutdown / operation canceled (Galaxy) |
| ttnn sdpa group [bh_p150b_civ2] | [bpc](https://github.com/tenstorrent/tt-metal/actions/runs/28087986929) | Pre-existing / no baseline | Compile error in `sparse_sdpa_writer.cpp` (file untouched by this branch); job absent on main's matrix |
| core ttnn unit test group [bh_p150b_civ2] | [bpc](https://github.com/tenstorrent/tt-metal/actions/runs/28087986929) | Infra / no baseline | DMA page-pin `Invalid argument` (umd); job absent on main's matrix |
| Llama 3.2-3B unit [wh_n150] | [all-models ](https://github.com/tenstorrent/tt-metal/actions/runs/28088002561)| Unverifiable | Logs dropped by GitHub  root cause unretrievable |

For safety I have now also started a nightly tt-metal L2 tests.
